fairlearn/reductions/_moments/moment.py

```diff
@@ -25,31 +24,50 @@ class Moment:
     def __init__(self):
         self.data_loaded = False
 
-    def load_data(self, X, y, **kwargs):
+    def load_data(self,
+                  X,
+                  y: pd.Series,
+                  *,
+                  sensitive_features: pd.Series = None):
         """Load a set of data for use by this object.
 
-        The keyword arguments can contain a :code:`sensitive_features` array.
-
-        :param X: The feature data
-        :type X: array
-
-        :param y: The true label data
-        :type y: array
+        Parameters
+        ----------
+        X : array
+            The feature array
+        y : pandas.Series
+            The label vector
+        sensitive_features : pandas.Series
+            The sensitive feature vector (default None)
         """
         assert self.data_loaded is False, \
             "data can be loaded only once"
-        self.X = X
+        assert isinstance(y, pd.Series)
```

Comment:
Can't seem to comment further on the existing thread:
![image](https://user-images.githubusercontent.com/10245648/102422839-7be5ee80-3fbc-11eb-98dd-10a852027117.png)

I have yet to see a good reason for using asserts in actual shipped code like this, although I'm ready to be convinced otherwise.

For me it's either
- This condition can happen and we should have proper try-except with a descriptive error message OR
- This condition can't happen, in which case we don't need the assert either.

Reply:
"Can't" is a very strong word, especially in Python.

I think this is partly because it's not clear to me what the public API of the `Moment` objects is supposed to be. I'm pretty sure this method is only used 'internally' (rather than externally by users), but I'm not certain. And while I agree that it's not going to give the prettiest of error messages, it is at least going to be clear as to what's wrong. Rather than blowing up with something more obscure later in the method - or worse, not blowing up for some time.

Reply:
I'm generally in agreement with @romanlutz's  observation, but I think that @riedgar-ms has a point. I have definitely used asserts for what's effectively "consistency checks of data structures". They are only effective [in the debug mode](https://docs.python.org/3/reference/simple_stmts.html#assert), so if somebody executes python with `-O`, they are ignored.

That said, we have places in the codebase where asserts should be probably replaced by exceptions.

Reply:
Also, in this case, I'm only asserting the PEP484 annotation :-)

Since Moments are not really user-facing (apart from the constructor), I'm going to leave it in, but I agree that things we expect users to see should be proper exceptions.

Reply:
Don't want to drag this further, but them only running in debug mode is another very good reason not to use them if you ask me...

Reply:
OK. I'll remove. It was mainly so that when putting this together I could be sure that all paths ending up here had been through input validation and reforming.